llvm/lib/Analysis/InstructionSimplify.cpp

```diff
@@ -4585,6 +4585,57 @@ static Value *simplifySelectWithFakeICmpEq(Value *CmpLHS, Value *CmpRHS,
                                Pred == ICmpInst::ICMP_EQ);
 }
 
+/// Try to simplify the Select instruction consisting of add/or/xor.
+static Value *simplifySpecificBitOp(Value *CmpLHS, Value *CmpRHS,
+                                    Value *TrueVal, Value *FalseVal) {
+  Value *X, *Y;
+  // (X & Y) == 0 ? X | Y : X ^ Y   --> X ^ Y
+  // (X & Y) == 0 ? X ^ Y : X | Y   --> X | Y
+  // (X & Y) != 0 ? X | Y : X ^ Y   --> X | Y
+  // (X & Y) != 0 ? X ^ Y : X | Y   --> X ^ Y
+  if (match(CmpLHS, m_And(m_Value(X), m_Value(Y))) && match(CmpRHS, m_Zero())) {
+    if ((match(TrueVal, m_c_Or(m_Specific(X), m_Specific(Y))) &&
+         match(FalseVal, m_c_Xor(m_Specific(X), m_Specific(Y)))) ||
+        (match(FalseVal, m_c_Or(m_Specific(X), m_Specific(Y))) &&
+         match(TrueVal, m_c_Xor(m_Specific(X), m_Specific(Y)))))
+      return FalseVal;
+  }
+  // (X | Y) == 0 ? X & Y : X ^ Y   --> X ^ Y
+  // (X | Y) == 0 ? X ^ Y : X & Y   --> X & Y
+  // (X | Y) != 0 ? X & Y : X ^ Y   --> X & Y
+  // (X | Y) != 0 ? X ^ Y : X & Y   --> X ^ Y
+  if (match(CmpLHS, m_Or(m_Value(X), m_Value(Y))) && match(CmpRHS, m_Zero())) {
+    if ((match(TrueVal, m_c_And(m_Specific(X), m_Specific(Y))) &&
+         match(FalseVal, m_c_Xor(m_Specific(X), m_Specific(Y)))) ||
+        (match(FalseVal, m_c_And(m_Specific(X), m_Specific(Y))) &&
+         match(TrueVal, m_c_Xor(m_Specific(X), m_Specific(Y)))))
+      return FalseVal;
+  }
+  // (X ^ Y) == 0 ? X | Y : X & Y   --> X & Y
+  // (X ^ Y) == 0 ? X & Y : X | Y   --> X | Y
+  // (X ^ Y) != 0 ? X | Y : X & Y   --> X | Y
+  // (X ^ Y) != 0 ? X & Y : X | Y   --> X & Y
+  if (match(CmpLHS, m_Xor(m_Value(X), m_Value(Y))) && match(CmpRHS, m_Zero())) {
+    if ((match(TrueVal, m_c_And(m_Specific(X), m_Specific(Y))) &&
+         match(FalseVal, m_c_Or(m_Specific(X), m_Specific(Y)))) ||
+        (match(FalseVal, m_c_And(m_Specific(X), m_Specific(Y))) &&
+         match(TrueVal, m_c_Or(m_Specific(X), m_Specific(Y)))))
+      return FalseVal;
+  }
+
+  // (X ^ Y) == 0 could be transformed to X == Y
+  // (X ^ Y) != 0 could be transformed to X != Y
+  if (match(CmpLHS, m_Value(X)) && match(CmpRHS, m_Value(Y))) {
+    if ((match(TrueVal, m_c_And(m_Specific(X), m_Specific(Y))) &&
+         match(FalseVal, m_c_Or(m_Specific(X), m_Specific(Y)))) ||
+        (match(FalseVal, m_c_And(m_Specific(X), m_Specific(Y))) &&
+         match(TrueVal, m_c_Or(m_Specific(X), m_Specific(Y)))))
+      return FalseVal;
+  }
```

Comment:
```suggestion
    if ((match(TrueVal, m_c_And(m_Specific(CmpLHS), m_Specific(CmpRHS))) &&
         match(FalseVal, m_c_Or(m_Specific(CmpLHS), m_Specific(CmpRHS)))) ||
        (match(FalseVal, m_c_And(m_Specific(CmpLHS), m_Specific(CmpRHS))) &&
         match(TrueVal, m_c_Or(m_Specific(CmpLHS), m_Specific(CmpRHS)))))
      return FalseVal;
```
These matches are no-op.

Reply:
Did you forget to fix it?

Reply:
I will change it to use `CmpLHS CmpRHS` to `X Y` instead using `match`  
ps. now I can use code review LOL